receiver/podmanreceiver/receiver.go

```diff
@@ -57,19 +69,90 @@ func newReceiver(
 		set:           set,
 	}
 
-	scrp, err := scraperhelper.NewScraper(typeStr, recv.scrape, scraperhelper.WithStart(recv.start))
+	return recv, err
+}
+
+func (r *receiver) registerMetricsConsumer(mc consumer.Metrics, set component.ReceiverCreateSettings) error {
+	r.metricsConsumer = mc
+	scrp, err := scraperhelper.NewScraper(typeStr, r.scrape)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return scraperhelper.NewScraperControllerReceiver(&recv.config.ScraperControllerSettings, set, nextConsumer, scraperhelper.AddScraper(scrp))
+	r.metricsComponent, err = scraperhelper.NewScraperControllerReceiver(&r.config.ScraperControllerSettings, set, mc, scraperhelper.AddScraper(scrp))
+	return err
 }
 
-func (r *receiver) start(context.Context, component.Host) error {
-	c, err := r.clientFactory(r.set.Logger, r.config)
-	if err == nil {
-		r.client = c
+func (r *receiver) registerLogsConsumer(lc consumer.Logs) {
+	r.logsConsumer = lc
+}
+
+func (r *receiver) Start(ctx context.Context, host component.Host) error {
+	// Check for logs pipeline
+	if r.logsConsumer == nil {
+		r.set.Logger.Debug("logs receiver is not set")
+	} else {
+		eventBackoff := backoff.NewExponentialBackOff()
+		eventBackoff.InitialInterval = 2 * time.Second
+		eventBackoff.MaxInterval = 10 * time.Minute
+		eventBackoff.Multiplier = 2
+		eventBackoff.MaxElapsedTime = 0
+
+		ctx, r.cancel = context.WithCancel(ctx)
+		backoffContext := backoff.WithContext(eventBackoff, ctx)
+
+		r.wg = &sync.WaitGroup{}
+		r.wg.Add(1)
+		go func() {
+			// Retry if any errors occur while getting the events.
+			defer r.wg.Done()
+			errorWhileRetry := backoff.Retry(func() error {
+				err := r.handleEvents(ctx, eventBackoff)
+				return fmt.Errorf("error fetching/processing events: %w", err)
+			}, backoffContext)
+			if errorWhileRetry != nil {
+				// context will be cancelled only in the case when shutdown will be called. So it's a graceful return.
+				if errors.Is(errorWhileRetry, context.Canceled) {
+					return
+				}
+				r.set.Logger.Error("retry failed", zap.Error(errorWhileRetry))
+				return
+			}
+		}()
 	}
-	return err
+
+	// Check for metrics pipeline
+	if r.metricsConsumer == nil {
+		r.set.Logger.Debug("metrics receiver is not set")
+	} else {
+		c, err := r.clientFactory(r.set.Logger, r.config)
+		if err != nil {
+			r.set.Logger.Error("error setting up client", zap.Error(err))
+		} else {
+			r.client = c
```

Comment:
Looks like duplicated across metrics and traces. Should be done once instead?

Reply:
Yes, the client can be shared. Made it common.